File: tools/vscode-punchcard/test/charCodeVectors.test.js
'use strict';

// Checks the committed cross-language parity fixtures
// (test/fixtures/char-code-*-vectors.csv) against the TypeScript codec that
// generated them. See test/fixtures/README.md for the file format this
// parses. A future Dart-side consumer compares the same files against
// lib/src/chars/char_code.dart (review finding VSC-5); until then this test
// guards against src/charCode.ts drifting from the committed fixture.

const assert = require('node:assert/strict');
const fs = require('node:fs');
const path = require('node:path');
const test = require('node:test');

const {
  bcdFromPunches,
  cardCodeFromPunches,
  glyphFromBcd,
  isGlyphColumn,
  machineSpecialName,
  punchesFromBcd,
} = require('../out/charCode.js');

const FIXTURES = path.join(__dirname, 'fixtures');
const PUNCH_VECTORS = path.join(FIXTURES, 'char-code-punch-vectors.csv');
const BCD_VECTORS = path.join(FIXTURES, 'char-code-bcd-vectors.csv');

/**
 * Parses one line of the fixture's format: every field double-quoted, `""`
 * escaping an internal quote, fields comma-separated.
 */
function parseCsvLine(line) {
  const fields = [];
  let i = 0;
  while (i < line.length) {
    if (line[i] !== '"') {
      throw new Error(`field ${fields.length} of ${JSON.stringify(line)} is not quoted`);
    }
    i++;
    let field = '';
    while (i < line.length) {
      if (line[i] === '"') {
        if (line[i + 1] === '"') {
          field += '"';
          i += 2;
        } else {
          i++;
          break;
        }
      } else {
        field += line[i];
        i++;
      }
    }
    fields.push(field);
    if (line[i] === ',') {
      i++;
    }
  }
  return fields;
}

/** Reads a fixture file into an array of `{ column: value }` row objects. */
function readCsv(filePath) {
  const text = fs.readFileSync(filePath, 'utf8');
  const lines = text.split('\n').filter((line) => line.length > 0);
  const header = parseCsvLine(lines[0]);
  return lines.slice(1).map((line) => {
    const values = parseCsvLine(line);
    assert.equal(values.length, header.length, `column count in ${JSON.stringify(line)}`);
    const row = {};
    header.forEach((name, i) => {
      row[name] = values[i];
    });
    return row;
  });
}

test('the committed punch-pattern vectors match the TypeScript codec', () => {
  const rows = readCsv(PUNCH_VECTORS);
  assert.equal(rows.length, 4096, 'run: npm run vectors');
  rows.forEach((row, i) => {
    const punches = Number(row.punches);
    assert.equal(punches, i, 'the fixture is sorted by ascending punches');
    const bcd = bcdFromPunches(punches);
    assert.equal(row.bcd, bcd === null ? '' : String(bcd), `punches ${punches}`);
    assert.equal(row.card_code, cardCodeFromPunches(punches), `punches ${punches}`);
    assert.equal(
      row.is_glyph_column,
      isGlyphColumn(punches) ? 'true' : 'false',
      `punches ${punches}`,
    );
  });
});

test('the committed BCD vectors match the TypeScript codec', () => {
  const rows = readCsv(BCD_VECTORS);
  assert.equal(rows.length, 64, 'run: npm run vectors');
  rows.forEach((row, i) => {
    const bcd = Number(row.bcd);
    assert.equal(bcd, i, 'the fixture is sorted by ascending bcd');
    const punches = punchesFromBcd(bcd);
    assert.equal(row.punches, punches === null ? '' : String(punches), `bcd ${bcd}`);
    assert.equal(row.glyph, glyphFromBcd(bcd) ?? '', `bcd ${bcd}`);
    assert.equal(row.machine_special, machineSpecialName(bcd) ?? '', `bcd ${bcd}`);
  });
});

test('the BCD vectors carry the literal comma glyph correctly quoted', () => {
  const rows = readCsv(BCD_VECTORS);
  const commaRow = rows.find((row) => row.glyph === ',');
  assert.ok(commaRow, 'no row has glyph ","');
  assert.equal(Number(commaRow.bcd), 59);
});

File: test/char_code_vectors_test.dart
import 'dart:io';

import 'package:comtran/comtran.dart';
import 'package:test/test.dart';

/// The cross-language parity gate for the character code (CI-01; VSC-5).
///
/// The VS Code extension carries a TypeScript port of
/// `lib/src/chars/char_code.dart`. Both codecs must agree on every punch
/// pattern and every BCD code. The committed fixtures
/// `tools/vscode-punchcard/test/fixtures/char-code-*-vectors.csv` are the
/// shared contract: the extension's `charCodeVectors.test.js` checks the
/// TypeScript side, and this test checks the Dart side, against the same
/// bytes. A change to either codec that is not reflected in a regenerated
/// fixture (`npm run vectors` in `tools/vscode-punchcard/`) fails one side.
///
/// The fixture format is documented in
/// `tools/vscode-punchcard/test/fixtures/README.md`: every field is
/// double-quoted, `""` escapes an internal quote, an empty field means the
/// source function returned null.

const String _fixtures = 'tools/vscode-punchcard/test/fixtures';

List<String> _parseCsvLine(String line) {
  final fields = <String>[];
  var i = 0;
  while (i < line.length) {
    if (line[i] != '"') {
      throw FormatException('field ${fields.length} of $line is not quoted');
    }
    i++;
    final field = StringBuffer();
    while (i < line.length) {
      if (line[i] == '"') {
        if (i + 1 < line.length && line[i + 1] == '"') {
          field.write('"');
          i += 2;
        } else {
          i++;
          break;
        }
      } else {
        field.write(line[i]);
        i++;
      }
    }
    fields.add(field.toString());
    if (i < line.length && line[i] == ',') {
      i++;
    }
  }
  return fields;
}

List<Map<String, String>> _readCsv(String path) {
  final List<String> lines = File(
    path,
  ).readAsLinesSync().where((String line) => line.isNotEmpty).toList();
  final List<String> header = _parseCsvLine(lines.first);
  return [
    for (final String line in lines.skip(1))
      Map.fromIterables(header, _parseCsvLine(line)),
  ];
}

void main() {
  test('the committed punch-pattern vectors match the Dart codec', () {
    final List<Map<String, String>> rows = _readCsv(
      '$_fixtures/char-code-punch-vectors.csv',
    );
    expect(rows, hasLength(4096), reason: 'run: npm run vectors');
    for (var i = 0; i < rows.length; i++) {
      final Map<String, String> row = rows[i];
      final int punches = int.parse(row['punches']!);
      expect(punches, i, reason: 'the fixture is sorted by ascending punches');
      final int? bcd = bcdFromPunches(punches);
      expect(row['bcd'], bcd?.toString() ?? '', reason: 'punches $punches');
      expect(
        row['card_code'],
        cardCodeFromPunches(punches),
        reason: 'punches $punches',
      );
      expect(
        row['is_glyph_column'],
        isGlyphColumn(punches) ? 'true' : 'false',
        reason: 'punches $punches',
      );
    }
  });

  test('the committed BCD vectors match the Dart codec', () {
    final List<Map<String, String>> rows = _readCsv(
      '$_fixtures/char-code-bcd-vectors.csv',
    );
    expect(rows, hasLength(64), reason: 'run: npm run vectors');
    for (var i = 0; i < rows.length; i++) {
      final Map<String, String> row = rows[i];
      final int bcd = int.parse(row['bcd']!);
      expect(bcd, i, reason: 'the fixture is sorted by ascending bcd');
      final int? punches = punchesFromBcd(bcd);
      expect(row['punches'], punches?.toString() ?? '', reason: 'bcd $bcd');
      expect(row['glyph'], glyphFromBcd(bcd) ?? '', reason: 'bcd $bcd');
      expect(
        row['machine_special'],
        machineSpecialName(bcd) ?? '',
        reason: 'bcd $bcd',
      );
    }
  });

  test('the BCD vectors carry the literal comma glyph correctly quoted', () {
    final List<Map<String, String>> rows = _readCsv(
      '$_fixtures/char-code-bcd-vectors.csv',
    );
    final Map<String, String> commaRow = rows.firstWhere(
      (Map<String, String> row) => row['glyph'] == ',',
      orElse: () => fail('no row has glyph ","'),
    );
    expect(int.parse(commaRow['bcd']!), 59);
  });
}
